python/paddle/distributed/auto_parallel/intermediate/tensor_parallel.py

```diff
@@ -0,0 +1,236 @@
+        # TODO(yaliu): fit pp
+        # Get local mesh for current pp.
+        assert "mp" in self.global_mesh.dim_names
+        if "pp" in self.global_mesh.dim_names:
+            assert (
+                self.global_mesh.get_dim_size("pp") == 1
+            ), "Not support pp with mp for now."
+            mesh = self.global_mesh.get_mesh_with_dim("pp")[0]
+        else:
+            mesh = self.global_mesh
+        assert len(mesh.shape) in [1, 2]
+        return mesh
+
+    def match_layer(self, name):
+        # Match the layer to a plan.
+        # Will return the plan if the layer hits one, otherwise return None.
+        for key, plan in self.parallelize_plan.items():
+            shard_weight = True
+            shard_bias = True
+            # Find some plan for specific parameter, such as
+            # "lm_head.weight": ColWiseParallel()
+            # Only support weight or bias.
+            if key.endswith(".weight"):
+                key = key.replace(".weight", "")
+                shard_bias = False
+            elif key.endswith(".bias"):
+                key = key.replace(".bias", "")
+                shard_weight = False
+            re_find = re.match(key, name)
+            if key == name or (re_find is not None and re_find.string == name):
+                return plan, shard_weight, shard_bias
+        return None, None, None
+
+    def tensor_parallelizer_fn(self, model):
+        if self.parallelize_plan is None:
+            return
+        for name, layer in model.named_sublayers():
+            if len(layer.sublayers()) == 0:
+                plan, shard_weight, shard_bias = self.match_layer(name)
+                if plan is not None:
+                    plan.apply(layer, self.get_mesh(), shard_weight, shard_bias)
+        return model
+
+
+def tensor_parallel(model, parallelize_plan=None, optimizer=None):
+    """
+    Tensor parallel.
+    :param model: paddle.nn.Layer, the model to be shard into tensor parallel.
+    :param parallelize_plan: Dict, the plan to shard the layer.
+    :param optimizer: paddle.optimizer.Optimizer, the optimizer.
+    :return:
+        model: model after sharding
+        optimizer: optimizer after sharding
+
+    NOTE: the plan should be a dict maps layer name or parameter name to a split_plan,
+    which will be used to split the layer or the parameter. The name can be written in regular format.
+
+    An example for the plan is:
+    ```
+    plan = {
+        "llama.embed_tokens": ColWiseParallel(),
+        "llama.layers.*.self_attn.q_proj": ColWiseParallel(),
+        "llama.layers.*.self_attn.k_proj": ColWiseParallel(),
+        "llama.layers.*.self_attn.v_proj": ColWiseParallel(),
+        "llama.layers.*.self_attn.o_proj": RowWiseParallel(),
+        "llama.layers.*.mlp.gate_proj": ColWiseParallel(),
+        "llama.layers.*.mlp.up_proj": ColWiseParallel(),
+        "llama.layers.*.mlp.down_proj": RowWiseParallel(),
+        "lm_head.weight": ColWiseParallel(),
+    }
```

Comment:
there is a TODO. once the API is fully determined, we will provide a complete example code, not only plan.